### PR TITLE
Add socket commands to reconfigure haproxy without reloading

### PR DESCRIPTION
**Pull request by aiharos**
This adds reconfiguring haproxy backend servers through stats socket without needing a reload.

The server "slots" for adding new endpoints are pre-allocated in the configuration file. Running out of slots or having too many empty ones does trigger a reload, and the increment by which the slots are added or removed is configurable through configmap.

Any other change in configuration behaves as before and triggers a reload.

### CODE REVIEW

**Comment by jcmoraisjr** [comment]
Thanks for supporting this ingress controller!

A few tweaks to your PR:

* People might want the old behaviour. Add a bool option (default to false) eg `backend-slots` to turn on and off this strategy
* Leave OnUpdate small and move the code and dependencies to a new file. Package controller is fine
* Do you need a parameter to the stats socket? If so leave the default as /var/run/haproxy-stats.sock, instead this could be the hardcoded value.
* Use Server instead of Srv in order to avoid ambiguity with Service
* Make golint happy and remove the `_` here: `for k, _ := range toRemoveEndpoints {`

**Comment by aiharos** [comment]
Thanks for the comments, glad to be able to help! Reducing reloads should be a good quality-of-life improvement.

I've added an option "dynamic-scaling" that can be used to get back to previous behavior of reloading after every OnUpdate() (still through the added code-path, just without issuing socket commands and addition of empty server slots).

The stats socket location doesn't need to be a configurable config-map parameter, but it makes sense to have it defined in a single place, to prevent template and code going out of sync.

File: pkg/controller/stats_socket.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		updatedConfig.BackendSlots = map[string]types.HAProxyBackendSlots{}
+
+		for _, backendName := range updKeys {
+			newBackend := types.HAProxyBackendSlots{}
+			newBackend.FullSlots = map[string]types.HAProxyBackendSlot{}
+			for i, endpoint := range updBackendsMap[backendName].Endpoints {
+				newBackend.FullSlots[fmt.Sprintf("%s:%s", endpoint.Address, endpoint.Port)] = types.HAProxyBackendSlot{
+					fmt.Sprintf("server%04d", i),
+					&endpoint,
+				}
```
Comment (aiharos) [inline pkg/controller/stats_socket.go]: It is possible to use `addr:port` if `DynamicScaling` is false, I'll change to that. You are correct that it's not feasible if `DynamicScaling` is enabled.

File: pkg/controller/stats_socket.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		_, ok := s2[k]
+		if !ok {
+			s3[k] = v
+		}
+	}
+	return s3
+}
+
+// populate template variables and optionally issue socket commands to reconfigure haproxy without reloading
+func reconfigureBackends(haproxy *HAProxyController, updatedConfig *types.ControllerConfig) {
```
Comment (aiharos) [inline pkg/controller/stats_socket.go]: I'll give it a go. ;)

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
...
+// Reload the backend if the configuration has changed
 func (haproxy *HAProxyController) Reload(data []byte) ([]byte, bool, error) {
-	if !haproxy.configChanged(data) {
-		return nil, false, nil
-	}
+	/*
+		if !haproxy.configChanged(data) {
+			return nil, false, nil
+		}
+	*/
```
Comment (aiharos) [inline pkg/controller/controller.go]: Indeed, I'll work in a `DeepEqual` even when `DynamicScaling` is false.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
 
 // OnUpdate regenerate the configuration file of the backend
 func (haproxy *HAProxyController) OnUpdate(cfg ingress.Configuration) ([]byte, error) {
-	data, err := haproxy.template.execute(newControllerConfig(&cfg, haproxy))
+	updatedConfig := newControllerConfig(&cfg, haproxy)
+
+	reconfigureBackends(haproxy, updatedConfig)
+
+	data, err := haproxy.template.execute(updatedConfig)
```
Comment (aiharos) [inline pkg/controller/controller.go]: There is a benefit to generating the configuration file even if `reloadRequired` is false - having the configuration in sync with in-memory configuration makes it easy to check what is currently active by just inspecting haproxy.cfg - so I would advise to keep this.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
 
 // OnUpdate regenerate the configuration file of the backend
 func (haproxy *HAProxyController) OnUpdate(cfg ingress.Configuration) ([]byte, error) {
-	data, err := haproxy.template.execute(newControllerConfig(&cfg, haproxy))
+	updatedConfig := newControllerConfig(&cfg, haproxy)
+
+	reconfigureBackends(haproxy, updatedConfig)
```
Comment (aiharos) [inline pkg/controller/controller.go]: I was on the fence about this too - I'll switch to the less coupled variant.

File: pkg/controller/config.go [hunk 6]
```diff
@@ -75,25 +75,28 @@ func newHAProxyConfig(haproxyController *HAProxyController) *types.HAProxyConfig
 			DefaultMaxSize: 1024,
```
Comment (aiharos) [inline pkg/controller/config.go]: Absolutely!

File: pkg/controller/config.go [hunk 7]
```diff
@@ -75,25 +75,28 @@ func newHAProxyConfig(haproxyController *HAProxyController) *types.HAProxyConfig
...
+		BackendCheckInterval:        "2s",
+		Forwardfor:                  "add",
+		MaxConn:                     2000,
+		HSTS:                        true,
+		HSTSMaxAge:                  "15768000",
+		HSTSIncludeSubdomains:       false,
+		HSTSPreload:                 false,
+		StatsPort:                   1936,
+		StatsAuth:                   "",
+		DynamicScaling:              true,
```
Comment (aiharos) [inline pkg/controller/config.go]: Alright

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
...
 	err := ioutil.WriteFile(haproxy.configFile, data, 0644)
 	if err != nil {
 		return nil, false, err
 	}
+
+	if !haproxy.ReloadRequired {
+		glog.Infoln("reload not required")
+		return nil, false, nil
+	}
+	glog.Infoln("reload is required")
```
Comment (aiharos) [inline pkg/controller/controller.go]: Noted

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
...
 	// TODO missing HAProxy validation before overwrite and try to reload
 	err := ioutil.WriteFile(haproxy.configFile, data, 0644)
 	if err != nil {
 		return nil, false, err
 	}
+
+	if !haproxy.ReloadRequired {
+		glog.Infoln("reload not required")
+		return nil, false, nil
+	}
```
Comment (aiharos) [inline pkg/controller/controller.go]: I'll see about making this more informative

File: pkg/controller/controller.go (hunk 4 above)
Comment (jcmoraisjr) [inline pkg/controller/controller.go]: Just to let you know Ingress now implement it's own equals before call `OnUpdate` and `Reload`  doesn't exist anymore. I can update Ingress right now (you'd need to merge or rebase) or I can update Ingress on top of your code and remove the `configChanged` part afterwards. I'd follow the latter but this definitely is your choice: update ingress now or on top of your changes?

**Comment by aiharos** [comment]
The addition of builtin Ingress equality before invoking `OnUpdate` seems like it makes things a bit easier in that we no longer need to worry about reloading when nothing changed and DynamicScaling is disabled, so the commits don't add `DeepEqual` for that case. 

In removing `Reload` we can move the code from it into `OnUpdate`, change `haproxy.reloadRequired` into just a local variable within `OnUpdate` and keep generating haproxy.cfg (since due to the change it should only be generated when the Ingress has changed).

Hopefully I've addressed the rest of the comments and haven't forgotten something.

File: pkg/controller/controller.go (hunk 4 above)
Comment (aiharos) [inline pkg/controller/controller.go]: I agree with the latter, hope that the set of updates based on your comments makes it straightforward.

**Comment by jcmoraisjr** [comment]
Thanks!

**Comment by jcmoraisjr** [comment]
@aiharos FYI just updated the readme file with [dynamic-scaling](https://github.com/jcmoraisjr/haproxy-ingress#dynamic-scaling) config. Please send a new PR if you want to change something.

**Comment by aiharos** [comment]
I forgot about the docs, thanks for updating - looks good!

**Pull request merged**